src/gui/dbsettings/DatabaseSettingsWidgetMasterKey.cpp

```diff
@@ -103,7 +100,9 @@ void DatabaseSettingsWidgetMasterKey::load(QSharedPointer<Database> db)
 
     setAdditionalKeyOptionsVisible(hasAdditionalKeys);
 
-    m_isDirty = isDirty;
+    connect(m_passwordEditWidget->findChild<QPushButton*>("removeButton"), SIGNAL(clicked()), SLOT(markDirty()));
+    connect(m_keyFileEditWidget->findChild<QPushButton*>("removeButton"), SIGNAL(clicked()), SLOT(markDirty()));
+    connect(m_yubiKeyEditWidget->findChild<QPushButton*>("removeButton"), SIGNAL(clicked()), SLOT(markDirty()));
```

Comment:
What about the change button? I'm also generally not sure if this approach works, because components are initialised lazily. A remove button may not exist until it's been shown for the first time. connect() won't complain, but it won't work either.

Reply:
You wrote this code 😉

Reply:
Did I? I guess it works then. 😛

Reply:
I tested it with my "all key types" test database. I don't like that the ui doesn't fully initialize because it creates runtime bugs like you mentioned (and the crash bug that is in 2.4.0). Can we fix that behavior?

Reply:
I don't think we can. That's Qt-internal stuff.

Reply:
We could try that.

Reply:
Actually, we cannot. setupUi is only available for widgets created with uic. I remember I had similar issues with initial vertical layout calculation, which I tried to solve in the updateSize() method. Now I also remember that I do initialise the remove and add pages (which is why this snippet should work), but not the dynamic widget. I'll try and see if I can make that work for that as well and if I find a better method than manipulating layout policies.

Reply:
Update: I found a solution. The reason was actually not that Qt didn't initialise the component but that we didn't do it unless the QStackedWidget changed its index. I have submitted an additional commit that fixes that and also resets the password field echo mode to bullets on database open (we had that, but there was a weird if statement around it that didn't make sense).